include/rocksdb/utilities/table_properties_collectors.h

```diff
@@ -84,4 +85,99 @@ std::shared_ptr<CompactOnDeletionCollectorFactory>
+// need-compaction when for the tiering use case, it observes, at least "D" data
+// entries that are already eligible to be placed on the last level but are not
+// yet on the last level. "D" is the compaction trigger threshold set via the
+// initial construction or later via `SetCompactionTrigger` API.
+// When tiering is disabled for a column family, no observation is made, the SST
+// file won't be marked as need-compaction either.
+class CompactForTieringCollectorFactory
+    : public TablePropertiesCollectorFactory {
+ public:
+  // @param enabled. Whether to enable the collectors to observe and mark files
+  // as need compaction.
+  // @param compaction_triggers. Mapping from column family id to compaction
+  // trigger. Not set it for a column family, or set the compaction trigger to 0
+  // effectively will only collect stats and write a user property in sst file
+  // Set it to something other than 0 will also mark the file as need compaction
+  // if the number of observed eligible entries is equal to or higher than the
+  // trigger.
+  // No observation will be made if a column family that does not enable
+  // tiering, regardless of the configured compaction trigger.
+  CompactForTieringCollectorFactory(
+      const std::map<uint32_t, size_t>& compaction_triggers, bool enabled);
+
+  ~CompactForTieringCollectorFactory() {}
+
+  TablePropertiesCollector* CreateTablePropertiesCollector(
+      TablePropertiesCollectorFactory::Context context) override;
+
+  // Update the value of existing compaction triggers or add compaction triggers
+  // for new column families.
+  void SetCompactionTriggers(
+      const std::map<uint32_t, size_t>& compaction_triggers) {
+    WriteLock _(&rwlock_);
+    for (const auto& [cf_id, compaction_trigger] : compaction_triggers) {
+      auto iter = compaction_triggers_.find(cf_id);
+      if (iter == compaction_triggers_.end()) {
+        compaction_triggers_.emplace(cf_id, compaction_trigger);
+      } else {
+        iter->second = compaction_trigger;
+      }
+    }
+  }
+
+  // Update the value of existing compaction trigger or add one for a new
+  // column family.
+  void SetCompactionTrigger(uint32_t cf_id, size_t compaction_trigger) {
+    WriteLock _(&rwlock_);
+    auto iter = compaction_triggers_.find(cf_id);
+    if (iter == compaction_triggers_.end()) {
+      compaction_triggers_.emplace(cf_id, compaction_trigger);
+    } else {
+      iter->second = compaction_trigger;
+    }
+  }
+
+  const std::map<uint32_t, size_t>& GetCompactionTriggers() const {
+    ReadLock _(&rwlock_);
+    return compaction_triggers_;
+  }
+
+  size_t GetCompactionTrigger(uint32_t cf_id) const {
+    ReadLock _(&rwlock_);
+    auto iter = compaction_triggers_.find(cf_id);
+    if (iter == compaction_triggers_.end()) {
+      return 0;
+    }
+    return iter->second;
+  }
+
+  // To support dynamically enable / disable all the collectors.
+  void SetEnabled(bool enabled) {
```

Comment:
I'm not a fan of a separate enablement flag distinct from a threshold/trigger value. When you set the threshold/trigger, you expect it to take effect. But that might not be true if you previously disabled the whole thing. It's essentially another issue of unnecessary and confounding redundancy (which I seem to be fighting against recently).

In the case of a trigger count, setting to max value should explicitly disable. That would ensure no discontinuity in the parameter space, though in this case where 0 is probably nonsensical, you could make a case for 0 meaning disabled, even if it's nearly the opposite of 1.

A potential complication of using max value is that SIZE_MAX can vary for 32-bit vs. 64-bit. Given the option serialization support, it's probably better to commit to uint32_t or uint64_t (or change to a fractional threshold).

Reply:
I have updated the threshold to be ratio wise, that turned out to have other benefits, like we can make 0 represent disabling the whole thing, and bigger than 1 represent to always just write the number of entries in user properties, but never mark a file.